### PR TITLE
Conviva integration V3 player SDK upgrade

### PR DESCRIPTION
**Pull request by luckygoyal-bitmovin**
This PR is for update of Conviva integration for Android to support Bitmovin V3 player SDKs.

Changes
- Conviva integration code supports V3 player SDK
- Test cases updated for V3 player SDK

### CODE REVIEW

**Comment by luckygoyal-bitmovin** [comment]
@stonko1994 , @jarhoax , adding you as reviewers but consider this as optional. The changes here are on top of the test cases PR which was merged today to migrate to use V3 player SDK. Would appreciate if you are able to find some time our of your sprint schedule for a quick review.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt [hunk 1]
```diff
@@ -22,6 +22,7 @@ class AdsTrackingTests : TestBase() {
 
         // load source and verify
         loadSource(activityScenario, DEFAULT_DASH_VOD_SOURCE)
+        Thread.sleep(2000)
```
Comment (stonko1994) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt:25]: Why do you need this sleep now? 🤔

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java [hunk 2]
```diff
@@ -289,11 +253,24 @@ public void resumeTracking() {
     }
     // endregion
 
-    private void customEvent(BitmovinPlayerEvent event) {
+    private void customEvent(PlayerEvent event) {
```
Comment (stonko1994) [inline conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:256]: I'm not 100% sure but I think there should be a super type `Event` which could be used here or?

File: ConvivaExampleApp/build.gradle [hunk 3]
```diff
@@ -35,7 +35,7 @@ dependencies {
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.1.0'
 
     implementation 'androidx.appcompat:appcompat:1.2.0'
-    implementation 'com.bitmovin.player:playercore:2.67.0'
+    implementation 'com.bitmovin.player:player:3.8.1'
```
Comment (jarhoax) [inline ConvivaExampleApp/build.gradle:38]: nit: We just released `3.9.0`

File: ConvivaExampleApp/src/androidTest/java/com/bitmovin/analytics/convivaanalyticsexample/ExampleInstrumentedTest.java [hunk 4]
```diff
@@ -18,9 +18,5 @@
 public class ExampleInstrumentedTest {
     @Test
     public void useAppContext() {
-        // Context of the app under test.
-        Context appContext = InstrumentationRegistry.getTargetContext();
-
-        assertEquals("com.bitmovin.analytics.convivaanalyticsexample", appContext.getPackageName());
```
Comment (jarhoax) [inline ConvivaExampleApp/src/androidTest/java/com/bitmovin/analytics/convivaanalyticsexample/ExampleInstrumentedTest.java:24]: Why not removing the whole test file? If you're going to add instrumentation tests, you will anyways create specific test files most likely.

File: ConvivaTestApp/build.gradle [hunk 5]
```diff
@@ -39,7 +39,7 @@ android {
 
 dependencies {
     implementation 'androidx.appcompat:appcompat:1.0.0'
-    implementation "com.bitmovin.player:playercore:2.67.0"
+    implementation "com.bitmovin.player:player:3.8.1"
```
Comment (jarhoax) [inline ConvivaTestApp/build.gradle:42]: same nit as at the top

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/TestBase.kt [hunk 6]
```diff
@@ -33,19 +34,19 @@ open class TestBase {
     ) as HashMap<String, Any>
     val CUSTOM_ERROR_MESSAGE = "CUSTOM_ERROR_MESSAGE"
 
-    val DEFAULT_DASH_VOD_SOURCE = SourceItem(DASHSource("https://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd"))
+    val DEFAULT_DASH_VOD_SOURCE = Source.create(SourceConfig("https://bitmovin-a.akamaihd.net/content/MI201109210084_1/mpds/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.mpd", SourceType.Dash))
```
Comment (jarhoax) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/TestBase.kt:37]: Cool that you are using `Source` here. For consistency I'd suggest either using `SourceConfig` here or also using `Source` in business logic above. Basically it doesn't matter 😉 just another nitpick.

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt [hunk 7]
```diff
@@ -27,31 +27,31 @@ class MainActivity : AppCompatActivity() {
         val intent = intent
         val autoPlay = intent.getBooleanExtra(AUTOPLAY_KEY, false)
         var vmapTagUrl = intent.getStringExtra(VMAP_KEY)
-        bitmovinPlaybackConfiguration = PlaybackConfiguration()
-        bitmovinPlaybackConfiguration.isAutoplayEnabled = autoPlay
+        bitmovinPlaybackConfig = PlaybackConfig()
+        bitmovinPlaybackConfig!!.isAutoplayEnabled = autoPlay
```
Comment (jarhoax) [inline ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt:31]: The force unwrap here (and following) should not be needed.

Even better would be:

```suggestion
        bitmovinPlaybackConfig = PlaybackConfig().apply {
        	isAutoplayEnabled = autoPlay
        }
```

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt [hunk 8]
```diff
@@ -27,31 +27,31 @@ class MainActivity : AppCompatActivity() {
...
 
         // Creating a new PlayerConfiguration
-        bitmovinPlayerConfiguration = PlayerConfiguration()
-        // Assign created SourceConfiguration to the PlayerConfiguration
-        bitmovinPlayerConfiguration.playbackConfiguration = bitmovinPlaybackConfiguration
+        bitmovinPlayerConfig = PlayerConfig()
+        bitmovinPlayerConfig!!.playbackConfig = bitmovinPlaybackConfig!!
```
Comment (jarhoax) [inline ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt:35]: The force unwrap here should not be needed.

Even better would be:

```suggestion
        bitmovinPlayerConfig = PlayerConfig().apply {
        	playbackConfig = bitmovinPlaybackConfig
        }
```

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt [hunk 9]
```diff
@@ -27,31 +27,31 @@ class MainActivity : AppCompatActivity() {
...
             // Setup ad
             val vmapAdRoll = AdItem("", vmapAdSource)
 
             // Add the AdItems to the AdvertisingConfiguration
-            bitmovinAdConfiguration = AdvertisingConfiguration(vmapAdRoll)
+            bitmovinAdConfig = AdvertisingConfig(vmapAdRoll)
             // Assing the AdvertisingConfiguration to the PlayerConfiguration
             // All ads in the AdvertisingConfiguration will be scheduled automatically
-            bitmovinPlayerConfiguration.advertisingConfiguration = bitmovinAdConfiguration
+            bitmovinPlayerConfig!!.advertisingConfig = bitmovinAdConfig!!
```
Comment (jarhoax) [inline ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt:48]: Same pattern with `apply` applies here

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt (hunk 1 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt:25]: sleep() was earlier inside the activityScenario block in loadSource() method. When testing repeatedly with V3 upgrade, it turned out that the test results are not consistent. Moving the sleep outside of activityScenario blocks in all cases improved test result consistency, hence this change.

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt [hunk 10]
```diff
@@ -27,31 +27,31 @@ class MainActivity : AppCompatActivity() {
...
         }
 
         // Create new BitmovinPlayerView with our PlayerConfiguration
-        bitmovinPlayerView = BitmovinPlayerView(this, bitmovinPlayerConfiguration)
-        bitmovinPlayerView.layoutParams = LinearLayout.LayoutParams(
+        bitmovinPlayer = Player.create(this, bitmovinPlayerConfig!!)
+        bitmovinPlayerView = PlayerView(this, bitmovinPlayer)
+        bitmovinPlayerView!!.layoutParams = LinearLayout.LayoutParams(
```
Comment (jarhoax) [inline ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt:54]: Same pattern with `apply` applies here

File: build.gradle [hunk 11]
```diff
@@ -21,7 +21,7 @@ allprojects {
         google()
         jcenter()
         maven {
-            url 'http://bitmovin.bintray.com/maven'
+            url 'https://bitmovin.jfrog.io/artifactory/public-releases'
```
Comment (jarhoax) [inline build.gradle:24]: The correct url would be

```suggestion
            url 'https://artifacts.bitmovin.com/artifactory/public-releases'
```

File: conviva/build.gradle [hunk 12]
```diff
@@ -35,7 +38,8 @@ dependencies {
     androidTestImplementation 'androidx.test.ext:junit:1.1.1'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.1.0'
     api 'com.conviva.sdk:conviva-core-sdk:4.0.5'
-    implementation 'com.bitmovin.player:playercore:2.67.0'
+    implementation 'com.bitmovin.player:player:3.8.1'
```
Comment (jarhoax) [inline conviva/build.gradle:41]: `3.9.0` released 😏

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java (hunk 2 above)
Comment (jarhoax) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:256]: Yes there is the `Event` super type. Would reduce duplicated code for `PlayerEvent` and `SourceEvent`

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java [hunk 13]
```diff
@@ -358,9 +335,12 @@ private void updateSession() {
...
     private void createContentMetadata() {
-        SourceItem sourceItem = bitmovinPlayer.getConfig().getSourceItem();
-        if (sourceItem != null) {
-            contentMetadataBuilder.setAssetName(sourceItem.getTitle());
+        Source source = bitmovinPlayer.getSource();
+        if (source != null) {
+            SourceConfig sourceConfig = source.getConfig();
+            if (sourceConfig != null) {
+                contentMetadataBuilder.setAssetName(sourceConfig.getTitle());
+            }
```
Comment (jarhoax) [inline conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:343]: nit: you could use early return here to keep the happy path not indented.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt (hunk 1 above)
Comment (jarhoax) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt:25]: Without a deeper look that indicates something "wrong". Apart from that a sleep of 2 secs seems a bit high.

Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt:25]: You are right. Tests pass with even 1 second sleep but occasionally these failed with with shorter sleep specially on Android simulator. So I added 2 seconds for stability of test results. Do you foresee any concerns around keeping it 2 seconds?

File: ConvivaExampleApp/src/androidTest/java/com/bitmovin/analytics/convivaanalyticsexample/ExampleInstrumentedTest.java (hunk 4 above)
Comment (luckygoyal-bitmovin) [reply ConvivaExampleApp/src/androidTest/java/com/bitmovin/analytics/convivaanalyticsexample/ExampleInstrumentedTest.java:24]: Good idea. Will remove it.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt (hunk 1 above)
Comment (jarhoax) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt:25]: Hm, ok I can see why you need to sleep here (waiting for the `onActivity` callback).

However I'd still prefer leaving this sleep in the method where it is actually "needed" i.e. the `loadSource` call.

I also hear you on that this improves stability but usually something like this indicates async work, and maybe there is a better solution to it than putting 2 secs sleeps all over the place. 

> Do you foresee any concerns around keeping it 2 seconds?

basically no (apart from the mentioned already) but it will increase your test run time significantly. _Yes it has been like this before, but if you can fix/find the actual root cause it may increase your test run times a lot_

But as I do not know enough on your test setup and all the implications, I leave it up to you as the PR creator in the end on how you proceed with this.

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java (hunk 13 above)
Comment (luckygoyal-bitmovin) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:343]: The control is not returning if source or sourceConfig is null so cannot return early return here.

Comment (jarhoax) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:343]:
```Java
Source source = bitmovinPlayer.getSource();
if (source == null) return;

SourceConfig sourceConfig = source.getConfig();
if (sourceConfig == null) return;

...
```

Should work and improves readability imo.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt (hunk 1 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt]: moved Sleep call to load method

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java (hunk 2 above)
Comment (luckygoyal-bitmovin) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java]: updated

File: ConvivaExampleApp/build.gradle (hunk 3 above)
Comment (luckygoyal-bitmovin) [reply ConvivaExampleApp/build.gradle]: done

File: ConvivaExampleApp/src/androidTest/java/com/bitmovin/analytics/convivaanalyticsexample/ExampleInstrumentedTest.java (hunk 4 above)
Comment (luckygoyal-bitmovin) [reply ConvivaExampleApp/src/androidTest/java/com/bitmovin/analytics/convivaanalyticsexample/ExampleInstrumentedTest.java:24]: removed the file

File: ConvivaTestApp/build.gradle (hunk 5 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/build.gradle]: updated

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt (hunk 7 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt]: updated

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt (hunk 8 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt]: updated

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt (hunk 9 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt]: updated

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt (hunk 10 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt]: updated

File: build.gradle (hunk 11 above)
Comment (luckygoyal-bitmovin) [reply build.gradle:24]: updated

File: conviva/build.gradle (hunk 12 above)
Comment (luckygoyal-bitmovin) [reply conviva/build.gradle]: updated

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java (hunk 13 above)
Comment (luckygoyal-bitmovin) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:331]: The original code does not return if source of sourceConfig is null

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/TestBase.kt (hunk 6 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/TestBase.kt:37]: @jarhoax , can you please elaborate on this comment?

Comment (jarhoax) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/TestBase.kt:37]: Here you are using the new `Source` type (which has been introduced with v3), which wraps a `SourceConfig`. The latter has been there since v2 and has been updated to make migration as easy as possible. Somewhere in this PR (I couldn't find it anymore 👀 ) I thought you were operating/instantiating `SourceConfig`s instead of `Source`. While both is totally fine, I just asked for consistency here, so either go with `SouceConfig` (only) also in those tests or use `Source` in business logic as well. 

As I cannot find the place I was referring to initially, feel free to ignore the initial comment. As said anyways just a nitpick, and not a necessary change.

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java (hunk 13 above)
Comment (jarhoax) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:331]: I know, you don't have to return an instance or primitive, just break the execution order with `return;`.

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt [hunk 14]
```diff
@@ -27,34 +26,33 @@ class MainActivity : AppCompatActivity() {
...
+            bitmovinPlayerConfig.advertisingConfig = AdvertisingConfig(vmapAdRoll)
         }
 
         // Create new BitmovinPlayerView with our PlayerConfiguration
-        bitmovinPlayerView = BitmovinPlayerView(this, bitmovinPlayerConfiguration)
-        bitmovinPlayerView.layoutParams = LinearLayout.LayoutParams(
-            LinearLayout.LayoutParams.MATCH_PARENT,
-            LinearLayout.LayoutParams.MATCH_PARENT
-        )
+        bitmovinPlayer = Player.create(this, bitmovinPlayerConfig!!)
```
Comment (jarhoax) [inline ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt:49]:
```suggestion
        bitmovinPlayer = Player.create(this, bitmovinPlayerConfig)
```

**Review by jarhoax — APPROVED** [review]
LGTM

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/PlaybackControlsTrackingTests.kt [hunk 15]
```diff
@@ -25,16 +25,14 @@ class PlaybackContorlsTrackingTests: TestBase() {
...
         verifyPlaying(activityScenario = activityScenario)
 
         // pause playback
         activityScenario.onActivity { activity: MainActivity ->
             clearMocks(playerStateManagerMock!!)
             activity.bitmovinPlayer.pause()
-            Thread.sleep(2000)
         }
-
+        Thread.sleep(2000)
```
Comment (dweinber) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/PlaybackControlsTrackingTests.kt:35]: Why do we need the `Thread.sleep(2000)`'s in here? Is there no better way? Is "it" (whatever we're waiting for) guaranteed to be available/done within 2000ms?

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/PlaybackErrorTrackingTests.kt [hunk 16]
```diff
@@ -17,8 +17,8 @@ class PlaybackErrorTrackingTests: TestBase() {
         // load invalid asset without initializing Conviva session explicitly
         activityScenario.onActivity { activity: MainActivity ->
             activity.bitmovinPlayer.load(INVALID_DASH_VOD_SOURCE)
-            Thread.sleep(2000)
         }
+        Thread.sleep(2000)
```
Comment (dweinber) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/PlaybackErrorTrackingTests.kt:21]: Same question as in `PlaybackControlsTrackingTests.kt` around the `Thread.sleep`.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/SessionTerminationTests.kt [hunk 17]
```diff
@@ -33,8 +32,8 @@ class SessionTerminationTests : TestBase() {
             clearMocks(playerStateManagerMock!!)
             clearMocks(clientMock!!)
             activity.bitmovinPlayer.seek(activity.bitmovinPlayer.duration - 1)
-            Thread.sleep(4000)
         }
+        Thread.sleep(4000)
```
Comment (dweinber) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/SessionTerminationTests.kt:36]: Same question as in the other test files. And why do we need 4000 instead of 2000ms here?

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java [hunk 18]
```diff
@@ -472,43 +442,74 @@ private void trackAdEnd() {
...
+        }
+    };
+
+    private final EventListener<PlayerEvent.Warning> onPlayerWarningListener = new EventListener<PlayerEvent.Warning>() {
+        @Override
+        public void onEvent(PlayerEvent.Warning warningEvent) {
+            Log.d(TAG, "[Player Event] Warning");
+            try {
+                ensureConvivaSessionIsCreatedAndInitialized();
```
Comment (dweinber) [inline conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:499]: Shouldn't this be `ConvivaAnalytics.this. ensureConvivaSessionIsCreatedAndInitialized();` as well? If it's not necessary, we might want to remove it in the `onSourceErrorListener`. Seems inconsistent.

**Review by dweinber — COMMENTED** [review]
Looks good overall! Only a few comments to clarify before I approve.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/PlaybackControlsTrackingTests.kt (hunk 15 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/PlaybackControlsTrackingTests.kt:35]: The wait is for the action like load(), play(), pause() etc. to be finished and expected event to be triggered by player, captured by Conviva integration code and expected Conviva API call executed. 

The value of 2 seconds was arrived at by multiple test runs. Some operations finish in lesser time but I used 2 seconds sleep in most cases as this resulted in stable test results and also does not add lot of delay to test case execution time.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/SessionTerminationTests.kt (hunk 17 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/SessionTerminationTests.kt:36]: This cases need to wait for seek to be completed and the playback to be finished so increased sleep to 4 seconds.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/PlaybackControlsTrackingTests.kt (hunk 15 above)
Comment (dweinber) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/PlaybackControlsTrackingTests.kt:35]: Can't we wait for one (or more) player events or so instead? But I just realized this very similar discussion above with Lorenz. We can keep the discussion there, I'll resolve this one here.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/SessionTerminationTests.kt (hunk 17 above)
Comment (dweinber) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/SessionTerminationTests.kt:36]: Can't we wait for the `Seeked` event or so instead to know when seeking completed?

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java (hunk 18 above)
Comment (dweinber) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:499]: Good point, any input for us here @jarhoax ?

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/SessionTerminationTests.kt (hunk 17 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/SessionTerminationTests.kt:36]: I explored this when writing tests. To achieve this the test case will need to register events listeners, then add a mechanism to block execution until the expected event is received and continue execution on receiving the expected event. I did not find a direct way to implement this event based delay/wait mechanism using the test framework.

Comment (dweinber) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/SessionTerminationTests.kt:36]: Understood, fine for me then.

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java (hunk 18 above)
Comment (luckygoyal-bitmovin) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:499]: Although there should be no ambiguity in names between anonymous event handler inner classes and outer class, but  using `ConvivaAnalytics.this.` sounds like a cleaner approach to prevent any ambiguity in future as well.

Comment (luckygoyal-bitmovin) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:499]: On second thoughts, to make code consistent, the ConvivaAnalytics.this. prefix will need to be added at many places in all event handler classes. I will rather remove it from a few places where it is used.

Comment (luckygoyal-bitmovin) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:499]: updated

**Review by dweinber — APPROVED** [review]
Nice 👍
